presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java

```diff
@@ -726,6 +757,17 @@ private Optional<FunctionNamespaceManager<? extends SqlFunction>> getServingFunc
         return Optional.ofNullable(functionNamespaceManagers.get(typeSignatureBase.getTypeName().getCatalogName()));
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    public SpecializedFunctionKey getSpecializedFunctionKey(Signature signature)
```

Comment:
calling out to BuiltInTypeAndFunctionNamepsaceManager to use that functionality for other namespace managers seems to violate some of the boundaries of the abstraction. Maybe we should extract that specialized function cache logic out of BuiltInTypeAndFunctionNamespaceManager if it needs to be common to other function namespace managers.